In [252]:
import requests
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import re
from datetime import datetime

In [3]:
pd.set_option('display.max_rows',None)

# PoetryDB

In [ ]:
with requests.session() as r:
    authors = r.get(r'https://poetrydb.org/author').json()['authors']
    poems = {}
    json = r.get(fr'https://poetrydb.org/author/{authors[0]}').json()

    poems = {'author': [],
        'title': [],
        'lines': [],
        'linecount' : []}
    

    for author in tqdm(authors):
        json = r.get(fr'https://poetrydb.org/author/{author}').json()
        for poem in json:
            poems['author'].append(poem['author'])
            poems['title'].append(poem['title'])
            poems['lines'].append(poem['lines'])
            poems['linecount'].append(poem['linecount'])

    poems_df = pd.DataFrame(poems)

poems_df


In [ ]:
poems_df.to_csv('poems.csv', mode='w', header = True, index=False)

In [ ]:
poems_df.author.to_csv('poems-authors.csv', header=True, index=False)

# Web Scrape

## Get Poets

In [9]:
with requests.session() as r:
    r.proxies['http'] = 'socks5h://localhost:9050'
    r.proxies['https'] = 'socks5h://localhost:9050'

    
    url = f'https://en.wikipedia.org/wiki/List_of_English-language_poets'
        
    wiki = r.get(url)
    soup = BeautifulSoup(wiki.text, 'html.parser')
    

        

In [10]:
poets_raw = pd.DataFrame()

for group in soup.find_all('div', class_= 'div-col'):
        p_worker = []
        temp_df = pd.DataFrame()
        
        for li in group.find_all('li'):
            p_worker.append(li.text)
        temp_df['poet'] = p_worker
        
        poets_raw = pd.concat([temp_df,poets_raw], axis=0)
       

In [12]:
poets_raw.shape

(3198, 1)

### Write poets_raw DataFrame to csv

In [13]:
poets_raw.to_csv('poets-raw.csv',mode='w',header=True, index=False)

### Reload poets_raw as poets and complete clean

In [14]:
poets = pd.read_csv('poets-raw.csv')

## Find poet nationalities

In [15]:
nationalities = {'A':'australia', 'B': 'barbados', 'Bo': 'bosnia', 'C': 'canada', 'Ch': 'chile', 'Cu':'cuba','D':'dominica','De':'denmark','E':'england','F':'france','G':'germany','Ga':'gambia','Gd':'granada','Gh':'ghana','Gr':'greece'
,'Gu':'guiana','Gy':'guernsey','HK':'hong_kong','In':'india','IoM':'isle_of_man','Is':'israel','Ir':'ireland','It':'italy','J':'jamaica','Je':'jersey','Jp':'japan','K':'kenya','L':'lebanon','M':'malta','Me':'mexico','Mo':'montserrat'
, 'Ne':'nepal','Nf':'newfoundland','Ni':'nigeria','NI':'northern_ireland','Nt':'netherlands','NZ':'new_zealand','P':'pakistan','Pa':'palestine','Ph':'phillipines','PI':'pitcairn_islands','RE':'russian_empire','S':'scotland','SA':'south_africa'
,'Se':'serbia','SL':'saint_lucia','SLe':'sierra_leone','SLk':'sri_lanka','So':'somalia','Sw':'sweden','T':'trinidad_and_tobago','US':'united_states','W':'wales','Z':'zimbabwe'}

In [16]:
poets['nation_key'] = poets.poet.apply(lambda x: x.replace('(','').replace(')','').replace(',','').split()[-1])
poets.nation_key.unique()

array(['US', 'C', 'E', 'A', 'SA', 'US/C', 'Ir', 'S', 'J/C', 'SL', 'E/IoM',
       'E/W', 'W/E', 'PI', 'W', 'C/US', 'Ir/S', 'S/E', 'E/A', 'SLa',
       'Ir/E', 'E/F', 'Ir/NI', 'E/Ir', 'HK', 'J', 'Ir/Z', 'NZ/A', '1653',
       'Ni', 'Ph', 'J/E', 'Se/C', 'In', 'E/US', 'E/C', 'Me/C', 'So/C',
       'C/E', 'A/E', 'NZ', 'C/Ch', 'India', 'SM', 'S/C', 'Bo/C', 'NI',
       'J/US', 'C/De', 'T', '1945', 'Gh', 'US/E', 'Je', 'Ch/C', 'Ir/C',
       'E/Nt', 'Sw', 'C/A', 'NI/E', 'Cu', 'E/S', 'Ga', 'T/C', 'SA/E',
       'Ne', 'W/F', 'S/SA', 'SLk/C', 'Ni/US', 'RE/US', 'NI/Ir', 'Gu/E',
       'A/S', 'W/US', 'SA/C', 'US/SA', 'Mo/E', 'Z/SA', 'Ir]', 'US/De',
       'K/Z', 'Ir/US', 'Ir/SLk', 'Z', 'US/Nt', 'SA/A', 'SA/US', 'SA/Is',
       'Gd/C', 'W/A', 'T/E', 'Ir/I', 'Is/C', 'Nf', 'G/US', 'Gy', 'S/A',
       'L/US', 'US/Jp', 'In/E', 'SA/Nt', 'Gu/C', 'Gu', 'S/US', 'C]',
       'P/W', 'In/US', 'US/A', 'Ir/A', 'T/S', 'SLe', 'Ir/F', 'E/G',
       'US/Ir', 'US/F', 'S/NZ', 'G/C', 'It', '1633', 'B', 'SA/F', 'IoM',


In [17]:
poets.reset_index(drop=True,inplace=True)
poets

poet nation_key
0                      Matthew Zapruder (born 1967, US)         US
1                      Marya Zaturenska (1902–1982, US)         US
2                            Robert Zend (1929–1985, C)          C
3                     Benjamin Zephaniah (born 1958, E)          E
4                        Komninos Zervos (born 1950, A)          A
5                          David Zieroth (born 1946, C)          C
6                            Rachel Zolf (born 1968, C)          C
7                         Daniel Zomparelli (living, C)          C
8                        Carolyn Zonailo (born 1947, C)          C
9                        Louis Zukofsky (1904–1978, US)         US
10                            Fay Zwicky (1933–2017, A)          A
11                            Jan Zwicky (born 1955, C)          C
12                     Makhosazana Xaba (born 1957, SA)         SA
13                       Mitsuye Yamada (born 1923, US)         US
14                        Leo Yankevich (1961–2018, US)         US
15                      Morgan Yasbincek (born 1964, A)          A
16                   J. Michael Yates (1938–2021, US/C)       US/C
17                          W. B. Yeats (1865–1939, Ir)         Ir
18                             Jean Yoon (born 1962, C)          C
19                        Marly Youmans (born 1953, US)         US
20                          Monica Youn (born 1971, US)         US
21                          Andrew Young (1885–1971, S)          S
22                       Augustus Young (born 1943, Ir)         Ir
23                          d'bi Young (born 1977, J/C)        J/C
24                         Douglas Young (1913–1973, S)          S
25                          Edward Young (1683–1765, E)          E
26                        Ellen Young (c. 1810–1872, A)          A
27                             Ian Young (born 1945, C)          C
28                          Kevin Young (born 1970, US)         US
29                     Marguerite Young (1908–1995, US)         US
30                     Miriam Waddington (1917–2004, C)          C
31                          Michael Wade (1944–2004, C)          C
32                          Sidney Wade (born 1951, US)         US
33                              Fred Wah (born 1939, C)          C
34                             John Wain (1925–1994, E)          E
35                           A. E. Waite (1857–1942, E)          E
36                         Nayyirah Waheed (living, US)         US
37                        Diane Wakoski (born 1937, US)         US
38                        Derek Walcott (1930–2017, SL)         SL
39                         Anne Waldman (born 1945, US)         US
40               George Waldron (1690 – c. 1730, E/IoM)      E/IoM
41                        Keith Waldrop (born 1932, US)         US
42                     Rosmarie Waldrop (born 1935, US)         US
43                          Arthur Waley (1889–1966, E)          E
44                         Alice Walker (born 1944, US)         US
45                      Margaret Walker (1915–1998, US)         US
46                            Rob Walker (born 1953, A)          A
47                       Bronwen Wallace (1945–1989, C)          C
48                  Chris Wallace-Crabbe (born 1934, A)          A
49                         Edmund Waller (1606–1687, E)          E
50                          Tom Walmsley (born 1948, C)          C
51                           Agnes Walsh (born 1950, C)          C
52                      Catherine Walsh (born 1964, Ir)         Ir
53                   David Waltner-Toews (born 1948, C)          C
54                          Ania Walwicz (1951–2020, A)          A
55                         Connie Wanek (born 1952, US)         US
56                    John Powell Ward (born 1937, E/W)        E/W
57                              Ned Ward (1667–1731, E)          E
58                          Sarah Wardle (born 1969, E)          E
59                Anna Laetitia

In [18]:
def check_nationality(poet):
    poet = poet.replace('(','').replace(')','').replace(',','')
    nation_key = poet.split()[-1]
    try:
        try:
            return nationalities[nation_key]
        except:
            nation_keys = nation_key.split('/')
            nations = []
            for key in nation_keys:
                nations.append(nationalities[key])
            return nations
    except:
        return np.nan

check_nationality('Matthew Zapruder (born 1967, US)')

'united_states'

In [19]:
check_nationality('J. Michael Yates (1938–2021, US/C)')

['united_states', 'canada']

In [20]:
poets['nationality'] = poets.poet.apply(check_nationality).copy()

poets['nationality_1'] = poets['nationality'].apply(lambda x: x[0] if type(x) == list else x)
poets['nationality_2'] = poets['nationality'].apply(lambda x: x[1] if type(x) == list else np.nan)

In [21]:
poets

poet nation_key  \
0                      Matthew Zapruder (born 1967, US)         US   
1                      Marya Zaturenska (1902–1982, US)         US   
2                            Robert Zend (1929–1985, C)          C   
3                     Benjamin Zephaniah (born 1958, E)          E   
4                        Komninos Zervos (born 1950, A)          A   
5                          David Zieroth (born 1946, C)          C   
6                            Rachel Zolf (born 1968, C)          C   
7                         Daniel Zomparelli (living, C)          C   
8                        Carolyn Zonailo (born 1947, C)          C   
9                        Louis Zukofsky (1904–1978, US)         US   
10                            Fay Zwicky (1933–2017, A)          A   
11                            Jan Zwicky (born 1955, C)          C   
12                     Makhosazana Xaba (born 1957, SA)         SA   
13                       Mitsuye Yamada (born 1923, US)         US   
14                        Leo Yankevich (1961–2018, US)         US   
15                      Morgan Yasbincek (born 1964, A)          A   
16                   J. Michael Yates (1938–2021, US/C)       US/C   
17                          W. B. Yeats (1865–1939, Ir)         Ir   
18                             Jean Yoon (born 1962, C)          C   
19                        Marly Youmans (born 1953, US)         US   
20                          Monica Youn (born 1971, US)         US   
21                          Andrew Young (1885–1971, S)          S   
22                       Augustus Young (born 1943, Ir)         Ir   
23                          d'bi Young (born 1977, J/C)        J/C   
24                         Douglas Young (1913–1973, S)          S   
25                          Edward Young (1683–1765, E)          E   
26                        Ellen Young (c. 1810–1872, A)          A   
27                             Ian Young (born 1945, C)          C   
28                          Kevin Young (born 1970, US)         US   
29                     Marguerite Young (1908–1995, US)         US   
30                     Miriam Waddington (1917–2004, C)          C   
31                          Michael Wade (1944–2004, C)          C   
32                          Sidney Wade (born 1951, US)         US   
33                              Fred Wah (born 1939, C)          C   
34                             John Wain (1925–1994, E)          E   
35                           A. E. Waite (1857–1942, E)          E   
36                         Nayyirah Waheed (living, US)         US   
37                        Diane Wakoski (born 1937, US)         US   
38                        Derek Walcott (1930–2017, SL)         SL   
39                         Anne Waldman (born 1945, US)         US   
40               George Waldron (1690 – c. 1730, E/IoM)      E/IoM   
41                        Keith Waldrop (born 1932, US)         US   
42                     Rosmarie Waldrop (born 1935, US)         US   
43                          Arthur Waley (1889–1966, E)          E   
44                         Alice Walker (born 1944, US)         US   
45                      Margaret Walker (1915–1998, US)         US   
46                            Rob Walker (born 1953, A)          A   
47                       Bronwen Wallace (1945–1989, C)          C   
48                  Chris Wallace-Crabbe (born 1934, A)          A   
49                         Edmund Waller (1606–1687, E)          E   
50                          Tom Walmsley (born 1948, C)          C   
51                           Agnes Walsh (born 1950, C)          C   
52                      Catherine Walsh (born 1964, Ir)         Ir   
53                   David Waltner-Toews (born 1948, C)          C   
54                          Ania Walwicz (1951–2020, A)          A   
55                         Connie Wanek (born 1952, US)         US   
56                    John Powell Ward (born 1937, E/W)      

There are 267 poets with dual nationality - I will drop these rows to allow only one column for nationality without null values.

In [22]:
poets.drop(poets[~poets.nationality_2.isnull()].index, inplace=True)
poets.drop('nationality_2', axis=1,inplace=True)
poets.reset_index(drop=True,inplace=True)


In [23]:
poets.drop(['nationality','nation_key'],axis=1,inplace=True)

In [24]:
poets.rename(columns={'nationality_1':'nationality'},inplace=True)

In [25]:
poets['nationality'].value_counts()

united_states          787
england                671
canada                 512
australia              351
scotland               208
ireland                126
south_africa            80
wales                   52
new_zealand             51
jamaica                 15
northern_ireland        14
nigeria                 12
india                   11
trinidad_and_tobago      8
ghana                    3
jersey                   2
guiana                   2
hong_kong                1
newfoundland             1
isle_of_man              1
barbados                 1
italy                    1
sierra_leone             1
guernsey                 1
gambia                   1
zimbabwe                 1
nepal                    1
cuba                     1
sweden                   1
saint_lucia              1
pitcairn_islands         1
phillipines              1
malta                    1
Name: nationality, dtype: int64

## Find poet dates

In [30]:
def find_dates(poet):
    try:
        try:
            pattern = re.compile(r'living')
            match = re.search(pattern, poet)
            return match[0]
        except:
            pattern = re.compile(r'\d{4}(\S\d{4})?')
            match = re.search(pattern, poet)
            return match[0]
    except: 
        return np.nan

In [31]:
poets['dates'] = poets.poet.apply(find_dates).copy()

In [33]:
poets.dates.isnull().sum()

11

#### Drop Null Values

In [35]:
poets.dropna(how='any',subset=['dates'],inplace=True)
poets.reset_index(drop=True, inplace=True)

poets

poet          nationality  \
0                      Matthew Zapruder (born 1967, US)        united_states   
1                      Marya Zaturenska (1902–1982, US)        united_states   
2                            Robert Zend (1929–1985, C)               canada   
3                     Benjamin Zephaniah (born 1958, E)              england   
4                        Komninos Zervos (born 1950, A)            australia   
5                          David Zieroth (born 1946, C)               canada   
6                            Rachel Zolf (born 1968, C)               canada   
7                         Daniel Zomparelli (living, C)               canada   
8                        Carolyn Zonailo (born 1947, C)               canada   
9                        Louis Zukofsky (1904–1978, US)        united_states   
10                            Fay Zwicky (1933–2017, A)            australia   
11                            Jan Zwicky (born 1955, C)               canada   
12                     Makhosazana Xaba (born 1957, SA)         south_africa   
13                       Mitsuye Yamada (born 1923, US)        united_states   
14                        Leo Yankevich (1961–2018, US)        united_states   
15                      Morgan Yasbincek (born 1964, A)            australia   
16                          W. B. Yeats (1865–1939, Ir)              ireland   
17                             Jean Yoon (born 1962, C)               canada   
18                        Marly Youmans (born 1953, US)        united_states   
19                          Monica Youn (born 1971, US)        united_states   
20                          Andrew Young (1885–1971, S)             scotland   
21                       Augustus Young (born 1943, Ir)              ireland   
22                         Douglas Young (1913–1973, S)             scotland   
23                          Edward Young (1683–1765, E)              england   
24                        Ellen Young (c. 1810–1872, A)            australia   
25                             Ian Young (born 1945, C)               canada   
26                          Kevin Young (born 1970, US)        united_states   
27                     Marguerite Young (1908–1995, US)        united_states   
28                     Miriam Waddington (1917–2004, C)               canada   
29                          Michael Wade (1944–2004, C)               canada   
30                          Sidney Wade (born 1951, US)        united_states   
31                              Fred Wah (born 1939, C)               canada   
32                             John Wain (1925–1994, E)              england   
33                           A. E. Waite (1857–1942, E)              england   
34                         Nayyirah Waheed (living, US)        united_states   
35                        Diane Wakoski (born 1937, US)        united_states   
36                        Derek Walcott (1930–2017, SL)          saint_lucia   
37                         Anne Waldman (born 1945, US)        united_states   
38                        Keith Waldrop (born 1932, US)        united_states   
39                     Rosmarie Waldrop (born 1935, US)        united_states   
40                          Arthur Waley (1889–1966, E)              england   
41                         Alice Walker (born 1944, US)        united_states   
42                      Margaret Walker (1915–1998, US)        united_states   
43                            Rob Walker (born 1953, A)            australia   
44                       Bronwen Wallace (1945–1989, C)               canada   
45                  Chris Wallace-Crabbe (born 1934, A)            australia   
46                         Edmund Waller (1606–1687, E)              england   
47                          Tom Walmsley (born 1948, C)               canada   
48                           Agnes Walsh (born 1950, C)               canada   
49                      Catherine Walsh (born 1964,

### Strip poets names of nationality and dates

In [36]:
poets.poet = poets.poet.apply(lambda x: re.search(r'[A-z]*[ .]{0,2}[A-z]*[. ]{0,2}[A-z]*', x)[0]).copy()

In [37]:
poets

poet          nationality      dates
0                   Matthew Zapruder         united_states       1967
1                   Marya Zaturenska         united_states  1902–1982
2                        Robert Zend                canada  1929–1985
3                 Benjamin Zephaniah               england       1958
4                    Komninos Zervos             australia       1950
5                      David Zieroth                canada       1946
6                        Rachel Zolf                canada       1968
7                  Daniel Zomparelli                canada     living
8                    Carolyn Zonailo                canada       1947
9                     Louis Zukofsky         united_states  1904–1978
10                        Fay Zwicky             australia  1933–2017
11                        Jan Zwicky                canada       1955
12                  Makhosazana Xaba          south_africa       1957
13                    Mitsuye Yamada         united_states       1923
14                     Leo Yankevich         united_states  1961–2018
15                  Morgan Yasbincek             australia       1964
16                        W. B. Yeats              ireland  1865–1939
17                         Jean Yoon                canada       1962
18                     Marly Youmans         united_states       1953
19                       Monica Youn         united_states       1971
20                      Andrew Young              scotland  1885–1971
21                    Augustus Young               ireland       1943
22                     Douglas Young              scotland  1913–1973
23                      Edward Young               england  1683–1765
24                       Ellen Young             australia  1810–1872
25                         Ian Young                canada       1945
26                       Kevin Young         united_states       1970
27                  Marguerite Young         united_states  1908–1995
28                 Miriam Waddington                canada  1917–2004
29                      Michael Wade                canada  1944–2004
30                       Sidney Wade         united_states       1951
31                          Fred Wah                canada       1939
32                         John Wain               england  1925–1994
33                        A. E. Waite              england  1857–1942
34                   Nayyirah Waheed         united_states     living
35                     Diane Wakoski         united_states       1937
36                     Derek Walcott           saint_lucia  1930–2017
37                      Anne Waldman         united_states       1945
38                     Keith Waldrop         united_states       1932
39                  Rosmarie Waldrop         united_states       1935
40                      Arthur Waley               england  1889–1966
41                      Alice Walker         united_states       1944
42                   Margaret Walker         united_states  1915–1998
43                        Rob Walker             australia       1953
44                   Bronwen Wallace                canada  1945–1989
45                      Chris Wallace            australia       1934
46                     Edmund Waller               england  1606–1687
47                      Tom Walmsley                canada       1948
48                       Agnes Walsh                canada       1950
49                   Catherine Walsh               ireland       1964
50                      David Waltner               canada       1948
51                      Ania Walwicz             australia  1951–2020
52                      Connie Wanek         united_states       1952
53                          Ned Ward               england  1667–1731
54                      Sarah Wardle               england       1969
55                        Emily Warn         united_states     living
56             Sylvia Townsend Warner      

### Save poets dataframe to csv

In [38]:
poets.to_csv('poets.csv',mode = 'w', header = True, index=False)

### Use wiki scrape to find sex based on gender pronouns

In [39]:
with requests.session() as r:
    sexes = []
    male_regex = r'(?<= )[Hh][ei][ms]?(?= )'
    female_regex = r'(?<= )[SsHh][h]?[e][r]?(?= )'

    for iteration, poet in tqdm(enumerate(poets.poet)):
        if iteration % 100 == 0:
            tqdm.write(f'Scraping {poet}')
        url = f'https://en.wikipedia.org/wiki/{poet}'
        wiki = r.get(url)
        soup = BeautifulSoup(wiki.text, 'html.parser')

        try:
            intro = soup.find_all('p', class_=False)[:3]
            check = 0

            for para in intro:
                text = para.text

                if re.search(male_regex,text):
                    sexes.append('male')
                    check = 1
                    break
                elif re.search(female_regex,text):
                    sexes.append('female')
                    check = 1
                    break
            
            if check == 0:
                sexes.append(np.nan)
        except:
            tqdm.write(f'investigate {poet}')
            sexes.append(np.nan)
    
    sexes = pd.Series(sexes)
        

0it [00:00, ?it/s]

Scraping Matthew Zapruder 


100it [00:28,  3.39it/s]

Scraping Dawud Wharnsby 


200it [00:55,  2.71it/s]

Scraping Mona Van Duyn


300it [01:23,  3.30it/s]

Scraping Charles Tompson 


400it [01:51,  3.74it/s]

Scraping Anna Seward 


500it [02:18,  3.99it/s]

Scraping William Somervile 


600it [02:45,  2.41it/s]

Scraping Bobbi Sykes 


700it [03:12,  3.76it/s]

Scraping Isaac Rosenberg 


800it [03:42,  2.68it/s]

Scraping Marcella Polain 


900it [04:21,  3.41it/s]

Scraping William Neill 


1000it [04:51,  3.58it/s]

Scraping David Manicom 


1101it [05:19,  4.33it/s]

Scraping M


1200it [05:48,  3.61it/s]

Scraping Daniel David Moses


1300it [06:16,  3.83it/s]

Scraping J. Patrick Lewis


1400it [06:45,  3.69it/s]

Scraping Barbara Kingsolver 


1500it [07:15,  2.49it/s]

Scraping Trevor Joyce 


1600it [07:42,  3.87it/s]

Scraping Jill Hellyer 


1700it [08:09,  3.77it/s]

Scraping Harry Howith 


1800it [08:39,  2.82it/s]

Scraping Phyllis Gotlieb 


1900it [09:08,  3.38it/s]

Scraping Michael Field 


2000it [09:38,  3.23it/s]

Scraping John Ennis 


2100it [10:06,  3.01it/s]

Scraping James Devaney 


2200it [10:36,  3.70it/s]

Scraping Stephen Dunn 


2300it [11:06,  2.89it/s]

Scraping Marilyn Chin 


2400it [11:34,  3.86it/s]

Scraping Alfred Corn 


2500it [12:03,  3.85it/s]

Scraping Djuna Barnes 


2600it [12:31,  2.99it/s]

Scraping Thomas Blacklock 


2700it [13:00,  3.01it/s]

Scraping Brian Brett 


2800it [13:30,  3.60it/s]

Scraping Toyin Adewale


2900it [14:00,  3.61it/s]

Scraping Cassandra Atherton 


2920it [14:06,  3.45it/s]


In [40]:
print(sexes.value_counts())
print(sexes.isnull().sum())

male      1577
female     842
dtype: int64
501


In [41]:
poets['sex'] = sexes

#### Drop Null Values

In [46]:
poets.dropna(how='any',subset=['sex'],inplace=True)

In [48]:
poets.to_csv('poets.csv',mode='w',header=True,index=False)

## Check poets dataframe to confirm good split of values

In [52]:
poets = pd.read_csv('poets.csv')

In [53]:
poets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2419 entries, 0 to 2418
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   poet         2419 non-null   object
 1   nationality  2411 non-null   object
 2   dates        2419 non-null   object
 3   sex          2419 non-null   object
dtypes: object(4)
memory usage: 75.7+ KB


### Convert sex to binary

1: Female

0: Male

In [55]:
poets.sex = poets.sex.apply(lambda x: 1 if x == 'female' else 0).copy()

### Convert dates column

For dates containing a range, find the midpoint and convert to float.

For poets with dates == living, convert to 2022.

Else convert date to float.

In [63]:
def convert_date(date_string):
    re_pattern = re.compile(r'(?<=\d{4})[\S]')

    if date_string == 'living':
        return float(2022)
    elif re.search(re_pattern,date_string):
        dates = re.split(re_pattern,date_string)
        return (float(dates[1]) + float(dates[0]))/2
    else:
        return float(date_string)

convert_date('1902–1982')



1942.0

In [65]:
poets.dates = poets.dates.apply(convert_date).copy()

### Clean up final null  values for nationality

In [70]:
poets[poets.nationality.isnull()]

poet nationality   dates  sex
105   Nissanka Wijeyeratne          NaN  1965.5    0
164         Lady Mary Wroth         NaN  1619.0    1
326         Lasana M. Sekou         NaN  1959.0    0
332            Vikram Seth          NaN  1952.0    0
487            Abhi Subedi          NaN  1945.0    0
1230           James Joyce          NaN  1911.5    0
1721   Christopher Dewdney          NaN  1951.0    0
2177         Edmund Bolton          NaN  1575.0    0

In [71]:
poets.nationality.unique()

array(['united_states', 'canada', 'england', 'australia', 'south_africa',
       'ireland', 'scotland', 'saint_lucia', 'pitcairn_islands', 'wales',
       nan, 'hong_kong', 'jamaica', 'nigeria', 'india', 'new_zealand',
       'trinidad_and_tobago', 'northern_ireland', 'ghana', 'jersey',
       'sweden', 'gambia', 'nepal', 'zimbabwe', 'newfoundland', 'guiana',
       'italy', 'barbados', 'isle_of_man', 'malta'], dtype=object)

In [72]:
poets.iloc[105,1] = 'sri_lanka'
poets.iloc[164,1] = 'england'
poets.iloc[326,1] = 'saint_martin'
poets.iloc[332,1] = 'india'
poets.iloc[487,1] = 'nepal'
poets.iloc[1230,1] = 'ireland'
poets.iloc[1721,1] = 'canada'
poets.iloc[2177,1] = 'england'

In [73]:
poets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2419 entries, 0 to 2418
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   poet         2419 non-null   object 
 1   nationality  2419 non-null   object 
 2   dates        2419 non-null   float64
 3   sex          2419 non-null   int64  
dtypes: float64(1), int64(1), object(2)
memory usage: 75.7+ KB


### Remove trailing whitespace from poet names

In [84]:
poets.poet = poets.poet.apply(lambda x: x.rstrip()).copy()

In [85]:
poets.to_csv('poets.csv',mode='w',header=True,index=False)

# PoetryFoundation

In [132]:
kp = pd.read_csv('kaggle-set/PoetryFoundationData.csv', index_col=0, header = 0, names = ['title','poem','poet','tags'])

In [133]:
kp.head()

title  \
0  \r\r\n                    Objects Used to Prop...   
1  \r\r\n                    The New Church\r\r\n...   
2  \r\r\n                    Look for Me\r\r\n   ...   
3  \r\r\n                    Wild Life\r\r\n     ...   
4  \r\r\n                    Umbrella\r\r\n      ...   

                                                poem              poet tags  
0  \r\r\nDog bone, stapler,\r\r\ncribbage board, ...  Michelle Menting  NaN  
1  \r\r\nThe old cupola glinted above the clouds,...     Lucia Cherciu  NaN  
2  \r\r\nLook for me under the hood\r\r\nof that ...        Ted Kooser  NaN  
3  \r\r\nBehind the silo, the Mother Rabbit\r\r\n...   Grace Cavalieri  NaN  
4  \r\r\nWhen I push your button\r\r\nyou fly off...      Connie Wanek  NaN

The PoetryFoundation database has 4 columns: `title`, `poem`, `poet`, `tags`.

Our target columns for this project are: `country_of_origin`, `sex` and `time_period`.

I need to use the poets in the PoetryFoundation dataset to identify this information.

My first step is it use the poets data scraped from wikipedia above to find the target labels for each poet.

In [134]:
kp.poet = kp.poet.apply(lambda x: x.rstrip().lstrip())

## Use Wikipedia Scraped Data

In [135]:
wiki_scrape = pd.read_csv('wiki-scrape/poets.csv')

In [136]:
wiki_scrape.head(10)

poet    nationality   dates  sex
0    Matthew Zapruder  united_states  1967.0    0
1    Marya Zaturenska  united_states  1942.0    1
2         Robert Zend         canada  1957.0    0
3  Benjamin Zephaniah        england  1958.0    0
4     Komninos Zervos      australia  1950.0    0
5       David Zieroth         canada  1946.0    0
6         Rachel Zolf         canada  1968.0    1
7   Daniel Zomparelli         canada  2022.0    0
8     Carolyn Zonailo         canada  1947.0    1
9      Louis Zukofsky  united_states  1941.0    0

In [137]:
poems = kp.join(wiki_scrape.set_index('poet'), on = 'poet', how = 'left')

In [138]:
print(f' Value Counts:')
print(poems.sex.value_counts(normalize = True))
print()
print(f' Nulls: {poems.sex.isnull().sum()}')

 Value Counts:
0.0    0.684932
1.0    0.315068
Name: sex, dtype: float64

 Nulls: 8642


There are still 8642 rows with null values for sex. For this reason, I will use a database of baby names to label the rest of the columns.

This method is hazardous, which is why I wanted to use my wiki_scrape above first; many female poets in history use male pseudonyms to get published, so this method may mislabel some poets. Also, the dataset I am using has been compiled from statistics from the following countries:

- US
- UK
- Canada
- Australia

This means the model may not be able to label names from other countries.

### Finish labeling sex column using names db and manual effort

In [139]:
names = pd.read_csv('kaggle-set/name_gender_dataset.csv')

In [140]:
male_names = []
female_names = []

for name in names.Name.unique():
    temp = names[names.Name == name].copy()
    temp.sort_values(by = 'Gender', inplace=True)
    temp.reset_index(drop=True,inplace=True)

    if temp.shape[0] == 1:
        if temp.iloc[0,1] == 'F':
            female_names.append(name)
        else:
            male_names.append(name)
    else:
        if temp.iloc[0,2] > temp.iloc[1,2]:
            female_names.append(name)
        else:
            male_names.append(name)

In [141]:
print(len(male_names))
print(len(female_names))

50989
82921


In [142]:
poems.loc[poems['sex'].isnull(), 'sex'] = poems[poems.sex.isnull()].poet.apply(lambda x: 0.0 if x.split(' ')[0] in male_names else 1.0 if x.split(' ')[0] in female_names else np.nan).copy()

In [143]:
print(f' Value Counts:')
print(poems.sex.value_counts(normalize = True))
print()
print(f' Nulls: {poems.sex.isnull().sum()}')

 Value Counts:
0.0    0.61394
1.0    0.38606
Name: sex, dtype: float64

 Nulls: 799


In [144]:
poems.sex.isnull().sum()

799

In [145]:
sx_labels = pd.read_csv('label-sex-complete.csv')
sx_labels.head()

poet     sex
0        Heid E. Erdrich  female
1  Indran Amirthanayagam    male
2        Osip Mandelstam    male
3   Sueyeun Juliette Lee  female
4    Sunnylyn Thibodeaux  female

In [146]:
poems.loc[poems.sex.isnull(),'sex'] = poems[poems.sex.isnull()].poet.apply(lambda x: sx_labels[sx_labels['poet'] == x].sex.values[0])

In [148]:
poems.to_csv('poems-with-targets.csv')

## Scrape wikipedia to get country and dates of remaining poets

In [113]:
remaining = poems[poems.dates.isnull()].poet.unique()

'T. S. Eliot' in remaining

True

In [96]:
soups = []
r = requests.session()
r.proxies['http'] = 'socks5h://localhost:9050'
r.proxies['https'] = 'socks5h://localhost:9050'

for index, poet in tqdm(enumerate(remaining)):
        url = f'https://en.wikipedia.org/wiki/{poet}'
        soup = BeautifulSoup(r.get(url).text, 'html.parser')
        soups.append(soup)



2524it [44:35,  1.06s/it]


In [115]:
poet_soup = pd.DataFrame(zip(remaining,soups), columns = ['poets','soup'])

In [116]:
poet_soup.head(1)

poets                                               soup
0  Michelle Menting  [html, \n, [\n, [\n, <meta charset="utf-8"/>, ...

In [121]:
poet_soup.drop(poet_soup.loc[poet_soup.poets == 'Michelle Menting'].index).head(1)

poets                                               soup
1  Lucia Cherciu  [html, \n, [\n, [\n, <meta charset="utf-8"/>, ...

#### Remove poets without wikipedia articles

In [122]:
for poet, soup in poet_soup.values:
    if soup.find('table', id = 'noarticletext'):
        poet_soup.drop(poet_soup.loc[poet_soup.poets == poet].index, inplace=True)

In [125]:
poet_soup.head(1)

poets                                               soup
2  Grace Cavalieri  [html, \n, [\n, [\n, <meta charset="utf-8"/>, ...

#### get poets info card

In [71]:
import unicodedata

def find_info_card(soup):
    try:
        infobox = soup.find('table', class_ = 'infobox vcard')
        infobox_labels = infobox.find_all('th', class_ = 'infobox-label')
        infobox_data = infobox.find_all('td', class_ = 'infobox-data')
        info_dict = {}

        for label, data in zip(infobox_labels,infobox_data):
            info_dict[unicodedata.normalize('NFKD',label.text)] = unicodedata.normalize('NFKD',data.text)

        return info_dict
    except:
        return np.nan
    

In [44]:
poet_soup['infocard'] = poet_soup.soup.apply(find_info_card)


In [246]:
poet_soup['dates'] = poet_soup.infocard.apply(find_dates_from_card)

In [115]:
def find_country_of_origin(card):
    try:
        if 'Nationality' in card.keys():
            return card['Nationality']
        elif 'Place of birth' in card.keys():
            return card['Place of birth']
        elif 'Born' in card.keys():
            if re.search(r'[A-z ,]*', card['Born']):
               return re.findall(r'([A-z ,]*)', card['Born'])
        else:
            return np.nan
    except:
        return np.nan
        

In [116]:
poet_soup.country_of_origin = poet_soup.infocard.apply(find_country_of_origin).copy()

In [117]:
poet_soup[['poets','country_of_origin','dates','infocard']]

poets                                  country_of_origin  \
0          Grace Cavalieri                                           American   
1             Simone White                                                NaN   
2          Heid E. Erdrich                   Turtle Mountain Ojibwe, American   
3              Sarah White                                                NaN   
4            Judith Harris                                                NaN   
...                    ...                                                ...   
1544          Lisa Barnett                                                NaN   
1545  Jeanne Murray Walker  [ , , , , , , , , , , , , ,  May , , , , , , ,...   
1546          Dylan Thomas  [Dylan Marlais Thomas, , , , , , , , , , , , ,...   
1547      John Wheelwright                                                NaN   
1548        Norman Macleod                                                NaN   

       dates                                           infocard  
0         []  {'Nationality': 'American', 'Alma mater': 'God...  
1       None                                                NaN  
2     1963.0  {'Born': ' (1963-11-26) November 26, 1963 (age...  
3       None                                                NaN  
4       None                                                NaN  
...      ...                                                ...  
1544    None                                                NaN  
1545  1944.0  {'Born': ' (1944-05-27) May 27, 1944 (age 77)P...  
1546  1933.5  {'Born': 'Dylan Marlais Thomas(1914-10-27)27 O...  
1547    None                                                NaN  
1548    None                                                NaN  

[1549 rows x 4 columns]

In [118]:
print(poet_soup.country_of_origin.isnull().sum())
print(poet_soup.dates.isnull().sum())
print(poet_soup.shape)

967
1001
(1549, 6)


In [119]:
def find_dates_from_card(card):
    try:
        dates = []
        if 'Born' in card.keys():
            dates.append(float(re.search(r'\d{4}',card['Born'])[0]))
        elif 'Date of birth' in card.keys():
           dates.append(float(re.search(r'\d{4}',card['Date of birth'])[0]))
        
        
        if 'Died' in card.keys():
            dates.append(float(re.search(r'\d{4}',card['Died'])[0]))
        elif 'Date of death' in card.keys():
            dates.append(float(re.search(r'\d{4}',card['Date of death'])[0]))
        
        if len(dates) == 1:
            return dates[0]
        elif len(dates) == 2:
            return sum(dates)/len(dates)
        else:
            return dates
    except:
        pass

In [60]:
poet_soup = pd.read_csv('poet-soup.csv')

In [62]:
poet_soup.soup = poet_soup.soup.apply(lambda x: BeautifulSoup(x, 'html.parser'))

In [72]:
poet_soup['infocard'] = poet_soup.soup.apply(find_info_card)


In [73]:
poet_soup.infocard

0       {'Nationality': 'American', 'Alma mater': 'God...
1                                                     NaN
2       {'Born': ' (1963-11-26) November 26, 1963 (age...
3                                                     NaN
4                                                     NaN
                              ...                        
1544                                                  NaN
1545    {'Born': ' (1944-05-27) May 27, 1944 (age 77)P...
1546    {'Born': 'Dylan Marlais Thomas(1914-10-27)27 O...
1547                                                  NaN
1548                                                  NaN
Name: infocard, Length: 1549, dtype: object

In [80]:
poet_soup.dates = poet_soup.infocard.apply(find_dates_from_card)

In [82]:
poet_soup[['poets','infocard','dates','country_of_origin']][~poet_soup.infocard.isnull()]

poets  \
0           Grace Cavalieri   
2           Heid E. Erdrich   
5               Truong Tran   
9             Eugene Gloria   
10    Indran Amirthanayagam   
...                     ...   
1532         Robert Marteau   
1538       Deirdre O'Connor   
1540            Al Maginnes   
1545   Jeanne Murray Walker   
1546           Dylan Thomas   

                                               infocard   dates  \
0     {'Nationality': 'American', 'Alma mater': 'God...      []   
2     {'Born': ' (1963-11-26) November 26, 1963 (age...  1963.0   
5     {'Born': 'Truong Tran1969Saigon, Vietnam', 'Oc...  1969.0   
9     {'Alma mater': 'San Francisco State University...      []   
10    {'Born': '1960Colombo', 'Genre': 'Poetry', 'No...  1960.0   
...                                                 ...     ...   
1532  {'Born': 'February 8th, 1925Virollet,  France'...  1968.0   
1538  {'Preceded by': 'Barry Maddern', 'Succeeded by...  1941.0   
1540  {'Position:': 'Tackle', 'Born:': '(1897-04-05)...      []   
1545  {'Born': ' (1944-05-27) May 27, 1944 (age 77)P...  1944.0   
1546  {'Born': 'Dylan Marlais Thomas(1914-10-27)27 O...  1933.5   

                                      country_of_origin  
0                                              American  
2                      Turtle Mountain Ojibwe, American  
5                                   Vietnamese American  
9                                                   NaN  
10                      ['', '', '', '', 'Colombo', '']  
...                                                 ...  
1532  ['February ', '', 'th, ', '', '', '', '', 'Vir...  
1538                                         Australian  
1540                                                NaN  
1545  [' ', '', '', '', '', '', '', '', '', '', '', ...  
1546  ['Dylan Marlais Thomas', '', '', '', '', '', '...  

[612 rows x 4 columns]

In [120]:
has_infocard = poet_soup[['poets','infocard','dates','country_of_origin']][~poet_soup.infocard.isnull()].copy()

In [121]:
has_infocard.reset_index(drop=True, inplace=True)

In [127]:
has_infocard.country_of_origin = has_infocard.country_of_origin.apply(lambda x: [word.strip() for word in x if word.strip() != ''] if type(x) == list else x).copy()

In [128]:
has_infocard.head()

poets                                           infocard  \
0        Grace Cavalieri  {'Nationality': 'American', 'Alma mater': 'God...   
1        Heid E. Erdrich  {'Born': ' (1963-11-26) November 26, 1963 (age...   
2            Truong Tran  {'Born': 'Truong Tran1969Saigon, Vietnam', 'Oc...   
3          Eugene Gloria  {'Alma mater': 'San Francisco State University...   
4  Indran Amirthanayagam  {'Born': '1960Colombo', 'Genre': 'Poetry', 'No...   

    dates                 country_of_origin  
0      []                          American  
1  1963.0  Turtle Mountain Ojibwe, American  
2  1969.0               Vietnamese American  
3      []                               NaN  
4  1960.0                         [Colombo]

In [149]:
poems.nationality.unique()


array([nan, 'united_states', 'england', 'trinidad_and_tobago',
       'hong_kong', 'canada', 'ireland', 'scotland', 'australia',
       'northern_ireland', 'wales', 'india', 'ghana', 'saint_lucia',
       'jamaica', 'new_zealand', 'barbados'], dtype=object)

In [156]:
has_infocard[['poets','country_of_origin']].to_csv('label_countries.csv')

In [157]:
labelled_countries = pd.read_csv('datasets/helper_sets/label_countries_final.csv')

In [161]:
poems.sex.isnull().sum()

0

In [169]:
poems.nationality.isnull().sum()

8642

In [172]:
poems.loc[poems.nationality.isnull(), 'nationality'] = poems.loc[poems.nationality.isnull()]['poet'].apply(lambda x: labelled_countries[labelled_countries.poets == x].country_of_origin.values[0] if x in labelled_countries.poets.unique() else np.nan).copy()

In [177]:
poems.loc[poems.dates.isnull(),'dates'] = poems.loc[poems.dates.isnull()]['poet'].apply(lambda x: has_infocard[has_infocard.poets == x].dates.values[0] if x in has_infocard.poets.unique() else np.nan).copy()

In [178]:
print(f'Missing nationalities: {poems.nationality.isnull().sum()}')
print(f'Missing sex: {poems.sex.isnull().sum()}')
print(f'Missing dates: {poems.dates.isnull().sum()}')

Missing nationalities: 5798
Missing sex: 0
Missing dates: 6059


In [193]:
poems.reset_index(drop=True,inplace=True)

In [206]:
poems['number_of_lines'] = poems.poem.apply(lambda x: len([word for word in x.splitlines() if word != ''])).copy()

In [208]:
poems.drop(poems.loc[poems['number_of_lines'] <= 1], axis = 1)

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[13898 rows x 0 columns]

In [217]:
poems.drop(poems.loc[poems['number_of_lines'] == 0].index, inplace=True)

In [222]:
poems.dates = poems.dates.apply(lambda x: np.nan if x == [] else x).copy()

In [230]:
poems.drop_duplicates(inplace=True)


title  \
6      \r\r\n                    Invisible Fish\r\r\n...   
7      \r\r\n                    Don’t Bother the Ear...   
9      \r\r\n                    ["Hour in which I co...   
80     \r\r\nfrom What the Heart Longs For When It On...   
112    \r\r\n                    Seventh Sphere (Satu...   
...                                                  ...   
13847  \r\r\n                    Incubus\r\r\n       ...   
13851  \r\r\n                    Descent\r\r\n       ...   
13852  \r\r\n                    Captain, Captive\r\r...   
13853  \r\r\n                    Apotheosis\r\r\n    ...   
13859  \r\r\n                    Address: The Archaea...   

                                                    poem  \
6      \r\r\nInvisible fish swim this ghost ocean now...   
7      \r\r\nDon’t bother the earth spirit who lives ...   
9      \r\r\nHour in which I consider hydrangea, a sa...   
80     \r\r\nWe spend the afternoon together watching...   
112    \r\r\nNo more hangings, no more gas chambers. ...   
...                                                  ...   
13847  \r\r\nThe chain uncouples, and his jacket hang...   
13851  \r\r\nMy father drummed darkness  Through the ...   
13852  \r\r\nOf your fate  Fast asleep  On the bed yo...   
13853  \r\r\nTaut with longing  You must become  The ...   
13859  \r\r\nAlthough most are totally naked  and too...   

                       poet  \
6                 Joy Harjo   
7                 Joy Harjo   
9              Simone White   
80     Sueyeun Juliette Lee   
112           Sun Yung Shin   
...                     ...   
13847          Craig Arnold   
13851        Samuel Menashe   
13852        Samuel Menashe   
13853        Samuel Menashe   
13859       Pattiann Rogers   

                                                    tags    nationality  \
6      Living,Time & Brevity,Relationships,Family & A...  united_states   
7      Religion,The Spiritual,Mythology & Folklore,Fa...  united_states   
9      Living,Parenthood,The Body,The Mind,Nature,Tre...            NaN   
80                            Animals,Photography & Film            NaN   
112                                   History & Politics            NaN   
...                                                  ...            ...   
13847  Living,Disappointment & Failure,Life Choices,M...  united_states   
13851                   Relationships,Family & Ancestors            NaN   
13852                                                NaN            NaN   
13853                                           Religion            NaN   
13859                           Arts & Sciences,Sciences            NaN   

        dates     sex  number_of_lines  
6      1951.0     1.0                1  
7      1951.0     1.0                1  
9         NaN     1.0                1  
80        NaN  female                1  
112       NaN     1.0                1  
...       ...     ...              ...  
13847  1988.0     0.0                1  
13851     NaN     0.0                1  
13852     NaN     0.0                1  
13853     NaN     0.0                1  
13859     NaN     1.0                1  

[3782 rows x 8 columns]

In [247]:
poems['number_of_stanzas'] = poems.poem.apply(lambda x: len([stanza for stanza in re.split(r'\r\r\n \r\r\n', x) if stanza != ''])).copy()

In [251]:
poems.head()

title  \
0  \r\r\n                    Objects Used to Prop...   
1  \r\r\n                    The New Church\r\r\n...   
2  \r\r\n                    Look for Me\r\r\n   ...   
3  \r\r\n                    Wild Life\r\r\n     ...   
4  \r\r\n                    Umbrella\r\r\n      ...   

                                                poem              poet tags  \
0  \r\r\nDog bone, stapler,\r\r\ncribbage board, ...  Michelle Menting  NaN   
1  \r\r\nThe old cupola glinted above the clouds,...     Lucia Cherciu  NaN   
2  \r\r\nLook for me under the hood\r\r\nof that ...        Ted Kooser  NaN   
3  \r\r\nBehind the silo, the Mother Rabbit\r\r\n...   Grace Cavalieri  NaN   
4  \r\r\nWhen I push your button\r\r\nyou fly off...      Connie Wanek  NaN   

     nationality   dates  sex  number_of_lines  number_of_stanzas  
0            NaN     NaN  1.0               20                  1  
1            NaN     NaN  1.0               14                  1  
2  united_states  1939.0  0.0               12                  1  
3  united_states     NaN  1.0               27                  1  
4  united_states  1952.0  1.0               28                  1

In [260]:
poems.to_csv(f'datasets/poems-with-targets-{now.strftime("%d-%m-%Y")}.csv',mode='w')

In [253]:
now = datetime.now()

In [254]:
now

datetime.datetime(2022, 2, 2, 9, 35, 21, 98670)

In [257]:
print(now.strftime('%d/%m/%Y'))

02/02/2022
